### PR TITLE
Add chi2 to calculate significance (still testing)

### PR DESCRIPTION
**Pull request by epelaaez**
IMPLEMENTATION
--------------------
I implemented the chi2 function as discussed and showed by Waly in Matlab. The code looks as follows;

```python
def get_prob(signal, idx):
    transform  = fft(signal[1]) # gets the FT of the values of the signal
    peak_value = transform[idx] # gets the FT coefficient for the detected frequency found by HFF

    real_std  = np.std(np.real(transform))
    imag_std  = np.std(np.imag(transform))
    avg_std   = (real_std + imag_std) / 2  # calculate deviation of real and imaginary part of the Ft and average them

    print(f"Real std: {real_std}. Imag std: {imag_std}")

    dof = 2
    x   = (np.abs(peak_value) ** 2) / (avg_std ** 2)
    chi =  chi2.cdf(x, dof)
    print(f"Prob: {chi}")
```

To test this out, I created a file called `chi_squared_testing.py` that creates signals with different error deviations, runs HFF to detect the most significant peak, and then uses `get_prob()` to get the probability of this being a real oscillation using chi2. The code for this testing looks as follows:

```python
test_stds = np.arange(0, 0.5, 0.1)
freq      = 5

for std in test_stds:
    print(f"Error deviation: {std}")
    signal    = create_data(m_phi=np.pi * ex_freq, deviation=std)
    idx, freq = get_hff(np.array(signal))
    print(f'Estimated frequency: {freq}') 
    get_prob(signal, idx)  
    print() # new line
```

The output for this particular configuration of `test_stds` is the following:
```
Error deviation: 0.0
Estimated frequency: 4.999999999999996
Real std: 2.461149352393693e-15. Imag std: 6.033462924309648
Prob: 0.9999999999997128

Error deviation: 0.1
Estimated frequency: 4.999999999999996
Real std: 2.2760857505677508. Imag std: 6.251457382820691
Prob: 0.9999923587817586

Error deviation: 0.2
Estimated frequency: 4.999999999999996
Real std: 4.4231738966372625. Imag std: 7.558441267850455
Prob: 0.9996020836235676

Error deviation: 0.30000000000000004
Estimated frequency: 148.99999999999986
Real std: 6.701113712753174. Imag std: 9.090799311035594
Prob: 0.08400340590006375

Error deviation: 0.4
Estimated frequency: 195.99999999999983
Real std: 8.951436638779686. Imag std: 9.16976493056563
Prob: 0.10222208224413089
```

As you can see with error deviations of 0 to 0.3, the HFF method is able to distinguish the expected frequency and the probability of it being right is very high; starting at almost 1 for error deviation of 0. Now, with 0.3 and 0.4, the HFF isn't even able to detect the frequency correctly, and the probability of the frequency detected (which is **wrong**) being right is very low, which is an indication that the function is working correctly.

An example where HFF is able to always detect the correct frequency is the following:

```
Error deviation: 0.2
Estimated frequency: 4.999999999999996
Real std: 4.716741049262739. Imag std: 7.247255042127496
Prob: 0.9935197093367082

Error deviation: 0.22
Estimated frequency: 4.999999999999996
Real std: 4.798113900624046. Imag std: 7.525876561661033
Prob: 0.9999833803589268

Error deviation: 0.24
Estimated frequency: 4.999999999999996
Real std: 5.387596217103451. Imag std: 7.982214194351032
Prob: 0.9761937081050643

Error deviation: 0.26
Estimated frequency: 4.999999999999996
Real std: 5.543046099031961. Imag std: 8.012152266044666
Prob: 0.9999021812338361

Error deviation: 0.27999999999999997
Estimated frequency: 4.999999999999996
Real std: 6.066314117845856. Imag std: 8.340683597121275
Prob: 0.9998302869628615
```

As you can see, the probability is always high since the correct frequency was detected. One important thing to note is that the probability doesn't necessarily decrease when the error deviation increases; I suspect this has to do with the error signal being random each run. Because of this randomness (I think), it may also happen that the probability is low even though the frequency is right. For example, I got the following after several runs of the code:

```
Error deviation: 0.2900000000000001
Estimated frequency: 4.999999999999996
Real std: 6.552957084453517. Imag std: 8.937406112439353
Prob: 0.3692179258169705
```

As you can see, the correct frequency was detected but the probability is low compared to the test ran above. However, I only got this after running the code many times. So I suspect this only happens when the random noise is generated in a very particular way.

TO DO
--------
- [x] Per Hugo and Daniel, implement the probability that no peak is higher than the peak we detected as most significant. The formula is: `(chi2.cdf(x, 2))^n`. Before doing this, we should make sure that the function works by itself.
- [x] Find if there are any cases in which the function doesn't work, i.e., gives a probability that doesn't make sense with the input signal.

### CODE REVIEW

**Comment by taejoonest** [comment]
Hey Emilio, why do you have to raise the exponent all the way to n? I am attempting to write the desired prob function, but does not cdf itself just give you that?

**Comment by waleey** [comment]
image
Here,  in the last two iteration, it gave wrong probability only because the estimated freqs were wrong. However, later I saw you got wrong probability even with correct frequency, and I couldn't find a explanation for it. My opinion is to use the Matlab for calculating chi-2 probability because I haven't observed any unreasonable behavior there. Also, since HFF sometimes misses the actual peak, we may associate an accuracy and confidence interval with it if necessary following simple statistics that I proposed.

**Comment by epelaaez** [comment]
@rabbit-hole-q I guess you're talking about the first TO DO item. We raise it to the `n` because we're considering all points and not the sole probability that the frequency detected is relevant. 

@waleey that sounds reasonable. Just one thing. I found this weird behavior after running the function many many times, therefore I suspect that it has to do with the randomness of the error. With real data, this shouldn't be much of a problem since we'll have many signals, and I suspect that a minority will have this weird behavior. 

Anyways, I think the best is that you look for this behavior on Matlab and if we don't find it, we should definitely "outsource" this job there. 

And about calculating general accuracy for HFF, that sounds good. I think the best is to get some signal to noise ratio upper boundary. So, for example, to say that HFF works until the noise is 0.5 times the real signal. Let me know if it sounds good.

**Comment by epelaaez** [comment]
An update on the weird behavior of the function.

I ran the code for error deviations in `test_stds = np.arange(0.1, 0.2, 0.01)` over 100 times and got the right frequency **and** reasonable frequency in all trials.

Then I tried for `test_stds = np.arange(0.2, 0.3, 0.01)`. After the 20th run, I saw a probability of 0.86, which isn't too low and can be attrtibuted to the noise. Around the 30th run, I saw a probability of 0.78, but the frequency was also wrongly detected, so no worries there. After the 70th run, I got a probability of 0.71 with the **correct** frequency. I don't know if we can attribute this to noise, but it certainly is something kind of weird. However, this appeared only after running the code 70 times, and didn't appear again in the last 30 runs; thus, I believe it has to do with the noise being generated in a very specific way this time. Most of the time in this 100 runs, the frequency was correctly detected with over 0.99 probability of being right. A few times, less than 10, the frequency was correctly detected with probability of ~0.85. 3 times the frequency was not correctly detected. And that one time, the correct frequency was detected with probability 0.71.

I finally tested with error deviations in `test_stds = np.arange(0.3, 0.4, 0.01)`. Right in the first 10 trials, only one detected the correct frequency; this is reasonable since the error deviations starts to get very close to the amplitude of our function. When the correct frequency was detected, the probability was 0.94. For the rest of incorrect frequencies, all probabilities were under 0.75. In the next 10 runs, HFF detected the correct frequency 3 times, for which the probability was 0.93 or higher. However, for one incorrect frequency it gave a probability of 0.85. I kept running this and for the most of correctectly detected frequencies, the probability was over 0.9, with some outliers around ~0.85. For the incorrect frequencies, the probability was mostly below 0.7, with few around 0.8, and one in 0.9.

In conclusion, this function indeed does have some weird behavior for certains signals. But I think this behavior depends on two things:
- The error deviation getting higher. When the error deviation is getting close to the ampltidue of the signal, it is very reasonable that it may create some trends that the chi squared function or HFF may detect as oscillatory. Even there, when the correct frequency was detected, most of the time the probability was around 0.9 or higher. 
- The randomness of the noise we are simulating. I suspect that sometimes the randomly generated noise may generate in a very particular way that can trick the chi squared or HFF functions. However, again, this only appeared a relatively small amount of times in all the testing I did.

**Comment by epelaaez** [comment]
I added some more testing files to test the whole functionality of what we've worked on. The `general_testing.py` file creates a csv file with all the testing data. I ran it only for the following parameters due to limited time, I'll try to run it for a bigger set of frequencies later.

```
test_stds  = np.arange(0, 0.6, 0.01)
test_freqs = np.arange(2, 100, 1)
```

The results for this run are available in the csv file I pushed.

Also feel free to run the test and analyze the data yourself.

**Note:** I found some weird behavior in the chi2 function. When the detected frequency is even, we need to divide by two the index of the frequency to get the accurate probability. I've added this modification to the `get_prob()` function directly, so it can be called just as before.

**Comment by epelaaez** [comment]
I implemented the uncertainty equation that Waly suggested, which is given by the following formula:
image
In the formula, `p` is the probability given by the chi squared function and `N` is the length of the Fourier transform. I added this to the `get_prob()` function, which now returns the probability and this uncertainty. I went for this since using the formula of `(chi2.cdf(x, 2))^n` took probabilities of around 0.9 and took them all the way to 0 for big samples.

I also ran the `general_test.py` file with this new uncertainty function and the following parameters:
```
test_stds     = np.arange(0, 0.5, 0.01)
test_freqs    = np.arange(1, 50, 0.5)
time_interval = 0.001
```

I've added the results in an csv file here. But they somewhat look like the following. The columns are, in order: real frequency, amplitude of the pure function, error deviation, detected frequency using HFF, chi2 probablity, uncertainty. As you can see, the first row detects the frequency correctly although the error deviation is higher than the amplitude, and the cdf gives a probability of 1 with uncertainty 0. And in the next row, the error becomes too high and the frequency is detected incorrectly, and the cdf probability and uncertainty make sense.
 
```
36.0, 0.037513179839879424, 0.04, 35.99999999999997, 1.0, 0.0
36.0, 0.037513179839879424, 0.05, 310.9999999999997, 0.43653046138418683, 0.015683482319558236
```

**Comment by epelaaez** [comment]
I'm going to merge this to get the code that generates the csv file. The chi2 file is most probably not the one we're going to use, but it may be good in the future for comparison.
